File: ovf/src/core/ViewFactor.cpp
#include "ovf/core/ViewFactor.hpp"

namespace openviewfactor {
  template <typename FLOAT_TYPE>
  ViewFactor<FLOAT_TYPE>::ViewFactor(std::vector<size_t> full_matrix_indices_map) {
    std::vector<FLOAT_TYPE> values;
    auto indices = full_matrix_indices_map;
    std::sort(indices.begin(), indices.end());
    for (int i=0; i < indices.size(); i++) {
      _nonzero_vf.push_back(std::make_pair(indices[i], values[i]));
    }
  }
  
  template <typename FLOAT_TYPE>
  ViewFactor<FLOAT_TYPE>& ViewFactor<FLOAT_TYPE>::link(std::shared_ptr<Triangulation<FLOAT_TYPE>> mesh) {
    if (_state != UNLINKED) {
      unlink();
    }
    _emitter = mesh;
    _state = LINKED_ONE_MESH;
    return *this;
  }
  template <typename FLOAT_TYPE>
  ViewFactor<FLOAT_TYPE>& ViewFactor<FLOAT_TYPE>::link(std::shared_ptr<Triangulation<FLOAT_TYPE>> emitter_mesh,
                                                       std::shared_ptr<Triangulation<FLOAT_TYPE>> receiver_mesh) {
    if (_state != UNLINKED) {
      unlink();
    }
    _emitter = emitter_mesh;
    _receiver = receiver_mesh;
    _state = LINKED_TWO_MESH;
    return *this;
  }
  template <typename FLOAT_TYPE>
  ViewFactor<FLOAT_TYPE>& ViewFactor<FLOAT_TYPE>::unlink() {
    if (_state != UNLINKED) {
      _emitter = nullptr;
      _receiver = nullptr;
      _state = UNLINKED;
    }
    return *this;
  }
  template <typename FLOAT_TYPE>
  bool ViewFactor<FLOAT_TYPE>::isLinked() const { return _state != UNLINKED; }

  template <typename FLOAT_TYPE>
  size_t ViewFactor<FLOAT_TYPE>::getNumEmitterElements() const {
    switch (_state) {
      case LINKED_ONE_MESH: //! fall through to LINKED_TWO_MESH
      case LINKED_TWO_MESH:
        return _emitter.size();
      default:
        throw std::exception("Cannot get mesh: No mesh linked to the result");
    }
  }
  template <typename FLOAT_TYPE>
  size_t ViewFactor<FLOAT_TYPE>::getNumReceiverElements() const {
    switch (_state) {
      case LINKED_ONE_MESH:
        return _emitter.size();
      case LINKED_TWO_MESH:
        return _receiver.size();
      default:
        throw std::exception("Cannot get mesh: No mesh linked to the result");
    }
  }

  template <typename FLOAT_TYPE>
  size_t ViewFactor<FLOAT_TYPE>::getNumMeshes() const { return static_cast<size_t>(_state); }
  template <typename FLOAT_TYPE>
  std::shared_ptr<Triangulation<FLOAT_TYPE>> ViewFactor<FLOAT_TYPE>::getEmitterMesh() const {
    switch (_state) {
      case LINKED_ONE_MESH: //! fall through to LINKED_TWO_MESH
      case LINKED_TWO_MESH:
        return _emitter;
      default:
        throw std::exception("Cannot get mesh: No mesh linked to the result");
    }
  }
  template <typename FLOAT_TYPE>
  std::shared_ptr<Triangulation<FLOAT_TYPE>> ViewFactor<FLOAT_TYPE>::getReceiverMesh() const {
    switch (_state) {
      case LINKED_ONE_MESH:
        return _emitter;
      case LINKED_TWO_MESH:
        return _receiver;
      default:
        throw std::exception("Cannot get mesh: No mesh linked to the result");
    }
  }
  template <typename FLOAT_TYPE>
  FLOAT_TYPE *ViewFactor<FLOAT_TYPE>::getViewFactorVector() const { return _nonzero_vf.data(); }

  template <typename FLOAT_TYPE>
  ViewFactor<FLOAT_TYPE>& ViewFactor<FLOAT_TYPE>::sortIndicesMap() {
    std::sort(_full_matrix_indices_map.begin(), _full_matrix_indices_map.end());
    std::sort(_nonzero_view_factors.begin(), _nonzero_view_factors.end(), _full_matrix_indices_map);
    return *this;
  }
  template <typename FLOAT_TYPE>
  ViewFactor<FLOAT_TYPE>& ViewFactor<FLOAT_TYPE>::setViewFactorVector(const std::vector<FLOAT_TYPE> &v) {
    _nonzero_view_factors = v;
    return *this;
  }
  template <typename FLOAT_TYPE>
  ViewFactor<FLOAT_TYPE>& ViewFactor<FLOAT_TYPE>::setElement(size_t emitter_index, size_t receiver_index, FLOAT_TYPE value) {
    auto full_matrix_index = (emitter_index * _receiver.size()) + receiver_index;
    std::vector<size_t>::iterator it = std::find(_full_matrix_indices_map.begin(), _full_matrix_indices_map.end(), full_matrix_index);
    if (it != _full_matrix_indices_map.end()) {
      size_t index_in_mapping_vector = it - _full_matrix_indices_map.begin();
      _nonzero_view_factors[index_in_mapping_vector] = value;
      return *this;
    } else {
      throw std::exception("Element attempting to be populated in ViewFactor::setElement was already determined to be zero and is therefore not present in the ViewFactor object.");
    }
  }

  template <typename FLOAT_TYPE>
  FLOAT_TYPE ViewFactor<FLOAT_TYPE>::getSurfaceToSurfaceAverageVF() const {
    //TODO FILL THIS OUT
    FLOAT_TYPE sum = 0.0;
    switch (_state) {
      case LINKED_ONE_MESH:
      case LINKED_TWO_MESH:
      default:
        throw std::exception("Cannot perform transformation: Result not linked to mesh");
    }
  }
  template <typename FLOAT_TYPE>
  FLOAT_TYPE ViewFactor<FLOAT_TYPE>::getEmitterElementToReceiverSurfaceVF(size_t emitter_index) const {
    //TODO FILL THIS OUT
    FLOAT_TYPE sum = 0.0;
    switch (_state) {
      case LINKED_ONE_MESH:
        break;
      case LINKED_TWO_MESH:
        break;
      default:
        throw std::exception("Cannot perform transformation: Result not linked to mesh");
    }
  }
  template <typename FLOAT_TYPE>
  FLOAT_TYPE ViewFactor<FLOAT_TYPE>::getReceiverElementFromEmitterSurfaceVF(size_t receiver_index) const {
    //TODO FILL THIS OUT
    FLOAT_TYPE sum = 0.0;
    switch (_state) {
      case LINKED_ONE_MESH:
        break;
      case LINKED_TWO_MESH:
        break;
      default:
        throw std::exception("Cannot perform transformation: Result not linked to mesh");
    }
  }
  template <typename FLOAT_TYPE>
  FLOAT_TYPE ViewFactor<FLOAT_TYPE>::getMatrixElementVF(size_t emitter_index, size_t receiver_index) const {
    return _nonzero_vf[(emitter_index * _receiver.size()) + receiver_index];
  }
}

File: movf/test/ViewFactor.m
classdef ViewFactor < handle
  properties (GetAccess = private, SetAccess = private)
    state ViewFactorState = ViewFactorState.UNLINKED
    emitter Mesh
    receiver Mesh
    nonzero_vf (1,:) IndexValuePair
  end

  methods
    function obj = ViewFactor(~)
    end

    function linkMeshes(obj, emitter_mesh, receiver_mesh)
      arguments
        obj ViewFactor
        emitter_mesh Mesh
        receiver_mesh Mesh
      end
      same_mesh = emitter_mesh == receiver_mesh;
      switch same_mesh
        case true
          obj.emitter = emitter_mesh;
          obj.state = ViewFactorState.LINKED_ONE_MESH;
        case false
          obj.emitter = emitter_mesh;
          obj.receiver = receiver_mesh;
          obj.state = ViewFactorState.LINKED_TWO_MESH;
      end
    end

    function state = getState(obj)
      state = obj.state;
    end
    function tf = isLinked(obj)
      tf = obj.state ~= ViewFactorState.UNLINKED;
    end

    function num_emitters = getNumEmitterElements(obj)
      if obj.isLinked
        num_emitters = obj.emitter.numElements;
      else
        exception = MException("ViewFactor:Unlinked", "Cannot get number of emitter elements, ViewFactor object is not linked to a mesh");
        throw(exception)
      end
    end
    function num_receivers = getNumReceiverElements(obj)
      switch obj.state
        case ViewFactorState.LINKED_ONE_MESH
          num_receivers = obj.getNumEmitterElements;
        case ViewFactorState.LINKED_TWO_MESH
          num_receivers = obj.receiver.numElements;
        case ViewFactorState.UNLINKED
          exception = MException("ViewFactor:Unlinked", "Cannot get number of receiver elements, ViewFactor object is not linked to a mesh");
          throw(exception)
      end
    end

    function num_meshes = getNumMeshes(obj)
      switch obj.state
        case ViewFactorState.LINKED_ONE_MESH
          num_meshes = 1;
        case ViewFactorState.LINKED_TWO_MESH
          num_meshes = 2;
        case ViewFactorState.UNLINKED
          exception = MException("ViewFactor:Unlinked", "Cannot get linked meshes, ViewFactor object is not linked to a mesh");
          throw(exception)
      end
    end

    function emitter_mesh = getEmitterMesh(obj)
      if obj.state ~= ViewFactorState.UNLINKED
        emitter_mesh = obj.emitter;
      else
        exception = MException("ViewFactor:Unlinked", "Cannot get the emitter mesh, ViewFactor object is not linked to a mesh");
        throw(exception)
      end
    end
    function receiver_mesh = getReceiverMesh(obj)
      switch obj.state
        case ViewFactorState.LINKED_ONE_MESH
          receiver_mesh = obj.emitter;
        case ViewFactorState.LINKED_TWO_MESH
          receiver_mesh = obj.receiver;
        case ViewFactorState.UNLINKED
          exception = MException("ViewFactor:Unlinked", "Cannot get the receiver mesh, ViewFactor object is not linked to a mesh");
          throw(exception)
      end
    end

    function setElement(obj, emitter_index, receiver_index, value)
      arguments
        obj ViewFactor
        emitter_index (1,1) uint64
        receiver_index (1,1) uint64
        value (1,1) double
      end
      full_matrix_index = ((emitter_index - 1) * obj.getNumReceiverElements) + (receiver_index - 1) + 1;
      pair = IndexValuePair(full_matrix_index, value);
      obj.nonzero_vf = [obj.nonzero_vf, pair];
    end

    function matrix_element = getMatrixElementVF(obj, emitter_index, receiver_index)
      arguments
        obj ViewFactor
        emitter_index (1,1) uint64
        receiver_index (1,1) uint64
      end
      full_matrix_index = ((emitter_index - 1) * obj.getNumReceiverElements) + (receiver_index - 1) + 1;
      map = obj.nonzero_vf;
      for map_index = 1 : length(map)
        pair = map(map_index);
        if full_matrix_index == pair.getFullMatrixIndex
          matrix_element = pair.getViewFactorValue;
        end
      end
    end

    function emitter_element_average = getEmitterElementToReceiverSurfaceVF(obj, emitter_index)
      arguments
        obj ViewFactor
        emitter_index (1,1) uint64
      end
      if obj.state == ViewFactorState.UNLINKED
        throw(MException('ViewFactor:AverageVF', 'Error getting emitting element-to-surface average: result is not linked a mesh'));
      end
      emitter_element_average = 0;
      for receiver_index = 1 : obj.getNumReceiverElements
        if emitter_index == receiver_index && obj.state == ViewFactorState.LINKED_ONE_MESH
          continue
        end
        emitter_element_average = emitter_element_average + obj.getMatrixElementVF(emitter_index, receiver_index);
      end
    end

    function receiver_element_average = getReceiverElementFromEmitterSurfaceVF(obj, receiver_index)
      arguments
        obj ViewFactor
        receiver_index (1,1) uint64
      end
      switch obj.state
        case ViewFactorState.LINKED_ONE_MESH
          receiver_element_average = obj.getEmitterElementToReceiverSurfaceVF(receiver_index);
        case ViewFactorState.LINKED_TWO_MESH
          receiver_element_average = 0;
          for emitter_index = 1 : obj.emitter.numElements
            area_reciprocal = obj.emitter.getElement(emitter_index).area / obj.receiver.getElement(receiver_index).area;
            receiver_element_average = receiver_element_average + obj.getMatrixElementVF(emitter_index, receiver_index) * area_reciprocal;
          end
        otherwise
          throw(MException('ViewFactor:AverageVF', 'Error getting receiving element-to-surface average: result is not linked a mesh'));
      end
    end

    function surface_to_surface_average = getSurfaceToSurfaceAverageVF(obj)
      if obj.state == ViewFactorState.UNLINKED
        throw(MException('ViewFactor:AverageVF', 'Error getting surface-to-surface average: result is not linked a mesh'));
      else
        surface_to_surface_average = 0;
        for emitter_index = 1 : obj.emitter.numElements
          surface_to_surface_average = surface_to_surface_average + obj.emitter.getElement(emitter_index).area * obj.getEmitterElementToReceiverSurfaceVF(emitter_index);
        end
        surface_to_surface_average = surface_to_surface_average / obj.emitter.area;
      end
    end
  end
end